채팅방 퇴장

JunBe opened the pull request:
## #️⃣ Issue Number

#87 

## 📝 작업 설명

변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요.   
어떻게보다 무엇을 왜 수정했는지 설명해주세요.

## 📸스크린샷 (선택)   

## 💬 리뷰 요구사항

리뷰어가 중점적으로 봐줬으면 좋겠는 부분이 있으면 적어주세요.   
논의해야할 부분이 있다면 적어주세요.   
ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요?

## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

JunBe committed:
feat: 채팅방 퇴장 기능 구현

- ChatRoom 엔티티에 구매자, 판매자의 퇴장 시간 저장 필드 추가
- 채팅방 나갈 시 퇴장 시간 업데이트 기능 추가

Resolves: #87

JunBe committed:
feat: 채팅방 퇴장 시 이후 메시지만 조회되도록 변경

Resolves: #87

JunBe committed:
feat: 채팅방 퇴장 후 채팅방 목록 조회 시 해당 채팅방 반환되지 않도록 변경

Resolves: #87

JinTaekLim commented on src/main/java/com/market/saessag/domain/chat/controller/ChatRoomController.java:
```diff
@@ -27,11 +28,18 @@ public ApiResponse<ChatRoomResponse> createChatRoom(@RequestBody ChatRoomRequest
+    @PostMapping("/left/{roomId}")
+    public ApiResponse<Void> leftChatRoom(@PathVariable Long roomId, HttpServletRequest request) {
+        chatRoomService.leftChatRoom(roomId, request);
+        return ApiResponse.success(SuccessCode.OK);
```
ApiResponse.success , ApiResponse.fail 과 같은 응답은 매개변수 없이 기본적으로 OK/FAIL 을 던지도록 변경하면 좋을 것같아요!

JinTaekLim commented on src/main/java/com/market/saessag/domain/chat/repository/ChatMessageRepository.java:
```diff
@@ -8,13 +8,23 @@
+            "AND ((m.sender.id = :userId AND (m.timeStamp > :leftAt OR :leftAt IS NULL))" +
+            "OR (m.sender.id <> :userId AND (m.timeStamp>:leftAt OR :leftAt IS NULL)))" +
+            "ORDER BY m.timeStamp DESC")
+    Page<ChatMessage> findMessagesAfterLeftTime(
```
현재는 leftAt 을 활용해 사용자가 나간 시점의 메세지 기록은 보이지 않도록 쿼리문을 작성하신 것같아요.

다만, LocalDateTime 을 기준으로 데이터를 조회하는 방법 이외의 다른 방법을 모색해보면 좋을 것같아요!


이전에 말씀드렸던 내용 중 ,
메세지를 기록하는 테이블에 상태 값 ( 메세지, 미디어, 입장/퇴장 안내문 ) 을 추가해 채팅방의 메세지/미디어/안내문을 사용자의 FK 값과 함께 저장해 특정 사용자의 상태 값이 [입장] 인 값을 기준으로 메세지를 구분하는 방법도 있을 것같습니다!

JinTaekLim commented on src/main/java/com/market/saessag/domain/chat/repository/ChatRoomRepository.java:
```diff
@@ -12,6 +14,13 @@ public interface ChatRoomRepository extends JpaRepository<ChatRoom, Long> {
+            "OR EXISTS (SELECT 1 FROM ChatMessage m WHERE m.chatRoom=r AND m.timeStamp>r.buyerLeftAt)))" +
+            "OR (r.seller=:user AND (r.sellerLeftAt IS NULL " +
+            "OR EXISTS (SELECT 1 FROM ChatMessage m WHERE m.chatRoom=r AND m.timeStamp>r.sellerLeftAt)))")
+    List<ChatRoom> findByBuyerOrSeller(@Param("user") User user);
```
해당 기능에 당장 위와같이 복잡한 쿼리는 불필요할 것같아요!
JPA로 구현이 되지 않는다면 테이블 구조를 변경하는 것도 고민해보시면 좋을 것같습니다.

JunBe replied:
현재 퇴장 후의 값만 필터 하여 보이게 하려다 보니 쿼리가 많이 복잡해지는 것 같네요.
나머지 남은 기능도 퇴장 후의 값들만 필터링 하려면 위 쿼리처럼 계속 작성해야 할 것 같아서 테이블 구조 다시 한번 생각해 보고 수정하겠습니다!

JinTaekLim commented on src/main/java/com/market/saessag/domain/chat/service/ChatRoomService.java:
```diff
@@ -91,4 +93,34 @@ private void subscribeUserToChatRoom(User user, ChatRoom chatRoom) {
+    }
+
+    // 세션에서 사용자 정보를 가져와서 검증
+    private User getUserFromSession(HttpServletRequest request) {
```
해당 메소드는 ChatRoomService 와는 맞지 않는 것같아요!

JunBe replied:
채팅방 서비스에 로그인 정보를 통해 유저 객체를 얻는 방식이 적절하지 않은 건가요?
아니면 사용자 정보를 검증하는 로직이 ChatRoomService에서 처리되고 있다는 점이 맞지 않은 건가요??

JinTaekLim replied:
사용자 정보 검증 로직이 ChatRoomService 에서 처리 되고 있는 부분입니다!

JunBe replied:
현재 유저 객체를 받아야 하는 메서드가 있는 서비스 로직 전부 getUserFromSession을 추가해서 검증하고 있었는데 그럼 검증 로직은 유저 서비스 쪽에서만 처리하고 유저 서비스에서 검증 메서드를 불러와서 사용해야 할까요??

JinTaekLim replied:
어떻게 조치할지에 관해서는 아현님과도 논의 후에 결정하면 좋을 것같아요!

지금처럼 각 Servcie 마다 동일한 역할의 메소드를 생성해 사용하게 되면, 이후 수정이 필요할 경우 모든 메소드를 찾아 수정해야하는 등 여러 문제점이 있겠지만, 결정적으로는 객체 지향에 적합한 방법은 아닌 것같습니다!

JunBe replied:
네 알겠습니다!

ahyeonkong replied:
RequestContextHolder를 이용해 세션 접근을 하도록 구성하면 좋을 것 같습니다!

JunBe replied:
넵 적용해보겠습니다!

JunBe committed:
refactor: 안 읽은 메시지 조회 방법 변경

- 기존 @Query를 이용하여 조인방식으로 조회하던 로직을 각 엔티티마다 JPA 기본 메서드를 사용해 개별로 조회하도록 변경
- 매개변수에 필요한 객체를 직접 전달하여 처리하도록 변경

Resolves: #87

JunBe committed:
style : WebSocket Controller 주석 추가

- 현재 로그인 한 사용자는 메시지 전송을 할 때 자신의 메시지는 읽음 처리가 되야함
- 현재 세션을 이용하고 있지 않기에 추후에 수정

Resolves: #87

JunBe committed:
refactor: 채팅방 반환 로직 변경

- @Query 방식 -> 개별 JPA로 조회하여 서비스 로직에서 필터하도록 변경

Resolves: #87

JunBe committed:
refactor: 메시지 검색 로직 변경

- 유저가 채팅방에서 마지막으로 떠난 이후의 메시지만 검색하도록 변경

Resolves: #87

JunBe committed:
refactor: 안 읽은 메시지 처리 로직 변경

- 안 읽은 메시지 조회, 카운트 마지막 퇴장 이후만 반환하도록 변경

Resolves: #87

JunBe committed:
refactor: 전체 메시지 조회 로직 변경

- JPA 사용하여 퇴장 이후의 전체 메시지 반환하도록 변경

Resolves: #87
